File: recover/src/recover/chaos.py
"""Chaos injectors for local testing. SQL copied verbatim from the original Rust chaos.rs."""
from __future__ import annotations

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncEngine


_STUCK_LEASES_SQL = text(
    """
    WITH picked AS (
        SELECT id FROM delivery_tasks
        WHERE status IN ('queued','retry_scheduled')
        ORDER BY random()
        LIMIT :count
        FOR UPDATE SKIP LOCKED
    )
    UPDATE delivery_tasks t
    SET status = 'sending',
        lease_token = gen_random_uuid()::text,
        lease_owner = 'chaos',
        lease_until = NOW() - INTERVAL '10 seconds',
        attempt_count = COALESCE(attempt_count, 0) + 1
    FROM picked
    WHERE t.id = picked.id
    """
)

_STUCK_OUTBOX_SQL = text(
    """
    WITH picked AS (
        SELECT id FROM outbox_events
        WHERE status = 'pending'
        ORDER BY random()
        LIMIT :count
        FOR UPDATE SKIP LOCKED
    )
    UPDATE outbox_events o
    SET status = 'publishing',
        locked_by = 'chaos',
        locked_until = NOW() - INTERVAL '30 seconds'
    FROM picked
    WHERE o.id = picked.id
    """
)

_ORPHAN_RETRY_SQL = text(
    """
    WITH picked AS (
        SELECT id FROM delivery_tasks
        WHERE status = 'retry_scheduled'
        ORDER BY random()
        LIMIT :count
        FOR UPDATE SKIP LOCKED
    )
    UPDATE delivery_tasks t
    SET available_at = NOW() - INTERVAL '5 minutes'
    FROM picked
    WHERE t.id = picked.id
    """
)


async def inject_stuck_leases(engine: AsyncEngine, count: int) -> int:
    async with engine.begin() as conn:
        res = await conn.execute(_STUCK_LEASES_SQL, {"count": count})
        return int(res.rowcount or 0)


async def inject_stuck_outbox(engine: AsyncEngine, count: int) -> int:
    async with engine.begin() as conn:
        res = await conn.execute(_STUCK_OUTBOX_SQL, {"count": count})
        return int(res.rowcount or 0)


async def inject_orphan_retry(engine: AsyncEngine, count: int) -> int:
    async with engine.begin() as conn:
        res = await conn.execute(_ORPHAN_RETRY_SQL, {"count": count})
        return int(res.rowcount or 0)

File: recover/src/chaos.rs
use anyhow::{Context, Result};
use sqlx::PgPool;

pub async fn inject_stuck_leases(pool: &PgPool, count: u32) -> Result<u64> {
    let res = sqlx::query(
        r#"
        WITH victims AS (
            SELECT id FROM delivery_tasks
            WHERE status IN ('queued','retry_scheduled')
              AND lease_until IS NULL
            ORDER BY random()
            LIMIT $1
            FOR UPDATE SKIP LOCKED
        )
        UPDATE delivery_tasks t
        SET status = 'sending',
            lease_owner = 'chaos-injector',
            lease_token = gen_random_uuid(),
            lease_until = NOW() - INTERVAL '1 minute',
            attempt_count = attempt_count + 1
        FROM victims
        WHERE t.id = victims.id
        "#,
    )
    .bind(count as i64)
    .execute(pool)
    .await
    .context("inject_stuck_leases")?;
    Ok(res.rows_affected())
}

pub async fn inject_stuck_outbox(pool: &PgPool, count: u32) -> Result<u64> {
    let res = sqlx::query(
        r#"
        WITH victims AS (
            SELECT id FROM outbox_events
            WHERE status = 'pending'
            ORDER BY random()
            LIMIT $1
            FOR UPDATE SKIP LOCKED
        )
        UPDATE outbox_events o
        SET status = 'publishing',
            locked_by = 'chaos-injector',
            locked_until = NOW() - INTERVAL '5 minutes'
        FROM victims
        WHERE o.id = victims.id
        "#,
    )
    .bind(count as i64)
    .execute(pool)
    .await
    .context("inject_stuck_outbox")?;
    Ok(res.rows_affected())
}

pub async fn inject_orphan_retry(pool: &PgPool, count: u32) -> Result<u64> {
    let res = sqlx::query(
        r#"
        WITH victims AS (
            SELECT id FROM delivery_tasks
            WHERE status = 'queued'
            ORDER BY random()
            LIMIT $1
            FOR UPDATE SKIP LOCKED
        )
        UPDATE delivery_tasks t
        SET status = 'retry_scheduled',
            available_at = NOW() - INTERVAL '10 minutes'
        FROM victims
        WHERE t.id = victims.id
        "#,
    )
    .bind(count as i64)
    .execute(pool)
    .await
    .context("inject_orphan_retry")?;
    Ok(res.rows_affected())
}
